File: Problem-3.cpp
// 648. Replace Words

// Time Complexity : O(n*k)[insertion] + (m*k)[checking] where k - length of words

// Space Complexity : O(n*k)

// Did this code successfully run on Leetcode : YES

// Appoarch: Tries - construct a trie and insert the words, check if each word can be replaces, 
// if yes then create a new string, add space and return.

#include <bits/stdc++.h>
using namespace std;

class Trie{
    public:
        bool isEnd;
        vector<Trie*> children; 
        Trie() : children(26,NULL), isEnd(false){}
};
class Solution {
    Trie* newNode;
public:
    Solution() : newNode(new Trie()) {}
    string check(string word){
        Trie* root = newNode;
        string s = "";
        for(auto i : word){
            if (root->isEnd) return s;  // Stop if a valid prefix is found
            if (!root->children[i - 'a']) break;
            s += i;
            root = root->children[i - 'a'];
        }
        return word;
    }
    void insert(string word) {
        Trie* curr = newNode;
        for(int i=0;i<word.length();i++){
            if(curr->children[word[i] - 'a'] == NULL){
                curr->children[word[i] - 'a'] = new Trie();
            }
            curr = curr->children[word[i] - 'a'];
        }
        curr->isEnd = true;
    }
    string replaceWords(vector<string>& dictionary, string sentence) {
        for(auto d:dictionary) insert(d);
        stringstream ss(sentence);
        string word = "",ans="";
        while(ss>>word){
            ans += check(word);
            ans += " ";
        }
        ans.pop_back();
        return ans;
    }
};

File: Problem_3.js
// Replace Words (https://leetcode.com/problems/replace-words/)

// TC: O(nk) n is the number of words in sentence; k is the length of longest word in dictionary
// SC: O(n) Size of trie
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach

class TrieNode {
    constructor() {
        this.children = new Array(26);
        this.isEnd = false;
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();
    }
    insert = function (word) {
        let curr = this.root;
        // Insert word in trie
        for (let i = 0; i < word.length; i++) {
            let ch = word[i];
            let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
            if (!curr.children[index]) {
                curr.children[index] = new TrieNode();
            }
            curr = curr.children[index]
        }
        curr.isEnd = true;
    }

    getSuccessor = function (word) {
        let curr = this.root;
        for (let i = 0; i < word.length; i++) {
            let ch = word[i];
            let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
            if (!curr.children[index]) {
                return word;
            }
            if (curr.children[index].isEnd) {
                return word.slice(0, i + 1);
            }
            curr = curr.children[index];
        }
        return word;
    }

}

/**
 * @param {string[]} dictionary
 * @param {string} sentence
 * @return {string}
 */
var replaceWords = function (dictionary, sentence) {
    if (dictionary === null || dictionary.length == 0)
        return sentence;
    if (sentence === null)
        return "";
    let trie = new Trie();
    for (let i = 0; i < dictionary.length; i++)
        trie.insert(dictionary[i]);

    let result = [];
    let words = sentence.split(/\s+/);
    for (let i = 0; i < words.length; i++) {
        let word = words[i];
        // Get successor if exists or return the word
        result.push(trie.getSuccessor(word));
    }

    return result.join(" ");
};